shark-hprof/src/main/java/shark/HprofDeobfuscator.kt

```diff
@@ -0,0 +1,239 @@
+package shark
+
+import shark.HprofRecord.HeapDumpEndRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ClassDumpRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ClassDumpRecord.FieldRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ClassDumpRecord.StaticFieldRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ClassSkipContentRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.InstanceDumpRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.InstanceSkipContentRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ObjectArrayDumpRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.ObjectArraySkipContentRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.PrimitiveArrayDumpRecord
+import shark.HprofRecord.HeapDumpRecord.ObjectRecord.PrimitiveArraySkipContentRecord
+import shark.HprofRecord.LoadClassRecord
+import shark.HprofRecord.StackFrameRecord
+import shark.HprofRecord.StringRecord
+import java.io.File
+
+/**
+ * Converts a Hprof file to another file with deobfuscated class and field names.
+ */
+class HprofDeobfuscator {
+
+  /**
+   * @see HprofDeobfuscator
+   */
+  fun deobfuscate(
+    proguardMappingFile: File,
+    inputHprofFile: File,
+    /**
+     * Optional output file. Defaults to a file in the same directory as [inputHprofFile], with
+     * the same name and "-deobfuscated" prepended before the ".hprof" extension. If the file extension
+     * is not ".hprof", then "-deobfuscated" is added at the end of the file.
+     */
+    outputHprofFile: File = File(
+        inputHprofFile.parent, inputHprofFile.name.replace(
+        ".hprof", "-deobfuscated.hprof"
+    ).let { if (it != inputHprofFile.name) it else inputHprofFile.name + "-deobfuscated" })
+  ): File {
+    val proguardMapping =
+      ProguardMappingReader(proguardMappingFile.inputStream()).readProguardMapping()
```

Comment:
This should probably be moved out of HprofDeobfuscator and into DeobfuscateHprofCommand, ie HprofDeobfuscator should be passed a ProguardMapping. Then tests become simpler (no need to parse a file, you can pass in a fake ProguardMapping.

Reply:
I created a helper class for creating `ProguardMapping` for testing purposes